File: keepr/Services/KeepsService.cs
namespace keepr.Services;

public class KeepsService
{
  private readonly KeepsRepository _keepsRepository;

  public KeepsService(KeepsRepository keepsRepository)
  {
    _keepsRepository = keepsRepository;
  }

  internal List<Keep> GetAllKeeps()
  {
    return _keepsRepository.Get();
  }

  internal Keep GetKeepById(int keepId)
  {
    Keep keep = _keepsRepository.GetById(keepId);
    if (keep == null)
    {
      throw new Exception("Keep does not exist");
    }
    keep.Views++;
    UpdateKeep(keep);
    return keep;
  }

  internal Keep CreateKeep(Keep newKeep)
  {
    return _keepsRepository.Create(newKeep);
  }

  internal Keep EditKeep(Keep keepData, string accountId)
  {
    Keep orig = GetKeepById(keepData.Id);
    if (orig.CreatorId != accountId)
    {
      throw new Exception("You can only edit your own keeps!");
    }
    orig.Name = keepData.Name ?? orig.Name;
    orig.Description = keepData.Description ?? orig.Description;
    orig.Img = keepData.Img ?? orig.Img;
    return _keepsRepository.Update(orig);
  }

  public void UpdateKeep(Keep k)
  {
    _keepsRepository.UpdateViews(k);
  }

  internal void RemoveKeep(int keepId, string accountId)
  {
    Keep keep = GetKeepById(keepId);
    if (keep.CreatorId != accountId)
    {
      throw new Exception("You can only delete your own keeps!");
    }
    _keepsRepository.Delete(keep);
  }
}

File: keepr.client/src/services/KeepsService.js
import { AppState } from "../AppState.js"
import { Keep } from "../models/Keep.js"
import Pop from "../utils/Pop.js"
import { api } from "./AxiosService.js"

class KeepsService {
  async getAllKeeps() {
    const res = await api.get("/api/keeps")
    AppState.keeps = res.data.map(k => new Keep(k))
  }

  async setActiveKeep(keepId) {
    AppState.activeKeep = null
    const keep = await this.getKeepById(keepId)
    AppState.activeKeep = keep
  }

  async createKeep(keepData) {
    const res = await api.post("/api/keeps", keepData)
    const keep = new Keep(res.data)
    AppState.keeps = [keep, ...AppState.keeps]
    AppState.profileKeeps = [keep, ...AppState.profileKeeps]
    AppState.activeKeep = null
    AppState.activeKeep = keep
  }

  async getProfileKeeps(profileId) {
    const res = await api.get(`/api/profiles/${profileId}/keeps`)
    AppState.profileKeeps = res.data.map(k => new Keep(k))
  }

  async getKeepById(keepId) {
    const res = await api.get(`/api/keeps/${keepId}`)
    const keep = new Keep(res.data)
    if (!keep) {
      Pop.error("Could not find keep")
    }
    return keep
  }

  async removeMyKeep(keepId) {
    const keep = await this.getKeepById(keepId)
    if (keep.creatorId != AppState.account.id) {
      Pop.error("You can only delete your own keeps")
    }
    await api.delete(`/api/keeps/${keep.id}`)
    AppState.profileKeeps = AppState.profileKeeps.filter(k => k.id != keep.id)
  }
}

export const keepsService = new KeepsService()